### ISSUE TITLE
CustomStorage.LogViewAdaptor could potentially cause an endless loop

### ISSUE THREAD

**Issue by rikbosch** [dotnet/orleans#5107]
When there is an exception during `ReadStateFromStorage` in` LogViewAdaptor.ReadAsync` which is unrecoverable,
the `JournalledGrain` ends up trying to load in an endless loop.

We would like to deactivate the grain (same behavior as InconsistentState exception).

I've tried deactivating the grain inside the `ICustomStorageInterface<T, object>.ReadStateFromStorage()` but that method is not run inside a grain context, but from a BatchWorker.

**Comment by sergeybykov** [comment]
@sebastianburckhardt What do you think?

**Comment by sebastianburckhardt** [comment]
PR #3773 contains some fixes to achieve a better retry logic, and also an API that lets you define your own error handling and diagnostics behavior for handling storage connection problems. Perhaps we could prioritize it.

**Comment by jamesbperry** [comment]
Glad to see that this is on the backlog. I just got bit by exactly what @roikonen observed in `ApplyUpdatesToStorage`'s lack of exponential backoff, and that led to a more fundamental question: handling non-transient write failures in a journaled grain.

**Scenario**:
1. [in some method of a journaled grain] `RaiseEvent(poisonEvent); await ConfirmEvents();`
2. [ApplyUpdatesToStorage] always `throw` or return `false` because the event is unacceptable for some reason

**Result**: A pegged CPU and a flooded log until the silo is restarted. 

There seems to be an assumption in `PrimaryBasedLogViewAdaptor` that any failures are either transient or due to resolvable inconsistency:
https://github.com/dotnet/orleans/blob/7720582a44723e3c8bb9ba3c371984b4b8f537be/src/Orleans.EventSourcing/Common/PrimaryBasedLogViewAdaptor.cs#L593-L595

(Side note: would this be a good place for the framework to `await LastPrimaryIssue.DelayBeforeRetry()` before looping?)

In the event of unresolved failures, @jason-bragg 's recommendation makes sense but feels at odds with the `PrimaryBasedLogViewAdaptor`'s logic:
> The safest path, imo, is to always deactivate the grain on storage operation (read, write, or clear) failure (of any kind). 
https://github.com/dotnet/orleans/issues/6128#issuecomment-555642545

So, in a simple scenario such as mine (no grain replication, re-entrancy, or unconfirmed events), is a viable strategy:
- The grain's `ApplyUpdatesToStorage` implementation uses a finite retry policy and catches any exceptions
- When the policy gives up for whatever reason, it flips a flag on the grain (say, `IsInconsistent`) but **reports success**, because
- Grain methods wanting to raise local events use a helper method which watches for the flag and destroys the grain
  ```c#
  private async Task TryWriteEvent(TEvent maybePoisonEvent)
  {
     RaiseEvent(maybePoisonEvent);
     await ConfirmEvents();
     if (IsInconsistent) throw new InconsistentStateException();
  }
  ```
- (optionally) A call filter traps the exception, e.g. if the domain model prefers a success/failure wrapper versus an exception
   ```c#
   try
    {
        await context.Invoke();
    }
    catch (InconsistentStateException)
    {
        // set the context.Response.Result
    }
   ```

It does appear that even if an incoming call filter catches the `InconsistentStateException`, the grain gets recycled by the runtime. `DeactivateOnIdle()` seemed like a viable alternative to throwing, but since the runtime goes to the trouble of handling `InconsistentStateException`s, that's probably the better option?

TL;DR - Any major flaws with letting the LogViewAdaptor temporarily think a write succeeded (so it stops retrying) and immediately burning the grain down afterwards?
